lucetc/src/function.rs

```diff
@@ -256,3 +263,75 @@ impl<'a> FuncEnvironment for FuncInfo<'a> {
         Ok(*pos.func.dfg.inst_results(inst).first().unwrap())
     }
 }
+
+///
+/// Writes a manifest of functions, with relocations, to the artifact.
+///
+/// THIS IS NOT CURRENTLY A COMPLETE LIST OF FUNCTIONS.
+///
+/// Currently the manifest returned here is the subset of functions
+/// that happen to have traps. This manifest is only used, right now,
+/// to look up trap manifest mappings, so this all lines up.
+///
+/// The *order* of these functions are not (currently) guaranteed either!
+/// Just that at the moment this order happens to be the same order as
+/// `FaerieTrapManifest` lists.
+///
+pub fn write_function_manifest(
+    manifest: &FaerieTrapManifest,
+    obj: &mut Artifact,
+) -> Result<Vec<FunctionSpec>, Error> {
+    let manifest_len_sym = "lucet_function_manifest_len";
+    obj.declare(&manifest_len_sym, Decl::data().global())
+        .context(format!("declaring {}", &manifest_len_sym))?;
+
+    let manifest_sym = "lucet_function_manifest";
+    obj.declare(&manifest_sym, Decl::data().global())
+        .context(format!("declaring {}", &manifest_sym))?;
+
+    let mut manifest_len_buf: Vec<u8> = Vec::new();
+    manifest_len_buf
+        .write_u32::<LittleEndian>(manifest.sinks.len() as u32)
+        .unwrap();
+    obj.define(manifest_len_sym, manifest_len_buf)
+        .context(format!("defining {}", &manifest_len_sym))?;
+
+    let mut manifest_buf: Cursor<Vec<u8>> = Cursor::new(Vec::with_capacity(
+        manifest.sinks.len() * size_of::<FunctionSpec>(),
+    ));
+
+    let mut function_manifest: Vec<FunctionSpec> = vec![];
+
+    /*
+     * !! This is brittle and should be suspect !!
+     *
+     * `.link()` is declare a relocation in the artifact
+     * relocating the bytes that HAPPEN to be the "addr"
+     * part of a FunctionSpec to point to the right function
+     *
+     * If `addr` moves in `FunctionSpec`, this will break!
+     */
+    for sink in manifest.sinks.iter() {
+        function_manifest.push(FunctionSpec::new(
+            0, // TODO: This is a lie. We don't have function address information at this point.
```

Comment:
Is there any kind of hint or initial value we can place here? I'd prefer to have a real address to the function, and a relocation to fix up when opened, but I think that also isn't exposed yet?

Reply:
I think `0` is fine as an initial value. We'll need testing to make sure the relocation ends up working correctly